### ISSUE TITLE
OUTPUT - INTO seems not to be working

### ISSUE THREAD

**Issue by saeohioalpha** [rails-sqlserver/activerecord-sqlserver-adapter#593]
I have an issue with output inserted on adapter v5.0.7 & Azure SQL (compat lvl 130)

The ActiveModel User.save method is triggering this. 

Not sure if this is my fault or if I am not putting a setting in that I need. 

This looks like it is trying to execute the following without the INTO section (from the log)
> EXEC sp_executesql N'INSERT INTO [users] ([Email], [PassHash], [First], [Last]) OUTPUT INSERTED.[id] VALUES (@0, @1, @2, @3)', N'@0 nvarchar(50), @1 nvarchar(50), @2 nvarchar(50), @3 nvarchar(50)', @0 = N'9ff019157811e18093fddb04daf3b69f@d.com', @1 = N'6935838143505C09CBB8779514089ED41233567F', @2 = N'sample', @3 = N'user'^[[0m  [["Email", nil], ["PassHash", nil], ["First", nil], ["Last", nil]]

I have tested this as working on the SQL server through ssms:
`DECLARE @inserted_id table (id int);
INSERT INTO [users] ([Email], [PassHash], [First], [Last]) 
	OUTPUT INSERTED.[id] AS INT INTO @inserted_id
	VALUES (
		N'9ff019157811e18093fddb04daf3b69f@d.com'
		,N'6935838143505C09CBB8779514089ED41233567F'
		,N'sample'
		,N'user'
	);
SELECT * FROM @Inserted_id;`

Also this works from the commandline using TSQL
`Setting example_db as default database in login packet
1> DECLARE @inserted_id table (id int);
2> INSERT INTO [users] ([Email], [PassHash], [First], [Last])
3>      OUTPUT INSERTED.[id] AS INT INTO @inserted_id
4>      VALUES (
5>              N'9ff019157811e18093fddb04daf3b69f@d.com'
6>              ,N'6935838143505C09CBB8779514089ED41233567F'
7>              ,N'sample'
8>              ,N'user'
9>      );
10> SELECT * FROM @Inserted_id;
11> GO
id
716162
(1 row affected)`

I also checked in the web console:
>  ActiveRecord::ConnectionAdapters::SQLServerAdapter.use_output_inserted
=> true

**Comment by metaskills** [comment]
Nothing here strikes me. Can you investigate more on what other settings might be happening on Azure that might make this not work. ANSI settings, etc? https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/blob/master/lib/active_record/connection_adapters/sqlserver_adapter.rb#L366

**Comment by saeohioalpha** [comment]
I haven't had much experience with this but shouldn't the EXEC sp_executesql statement in the log include the INTO section if it was working? How would the SQL server settings prevent that from being there in the rails app?

Attached is a screenshot of my DB settings:
![screen shot 2017-05-23 at 8 59 07 am](https://cloud.githubusercontent.com/assets/969559/26355314/2a2cd60a-3f96-11e7-9628-9e03b670a3f4.png)

**Comment by saeohioalpha** [comment]
I also had a funky issue about a week ago with the .exists method that would do a one as one query, but would have an offset fetch next without an order by. So maybe I do have something configured wrong. I couldnt get it fixed, I ended up just swapping exists with a where and checking myself.

**Comment by saeohioalpha** [comment]
Sorry I just realized I didn't include the Logged Error, the issue is that it is just using OUTPUT not OUTPUT INTO since I have a trigger on the table.

ActiveRecord::StatementInvalid (TinyTds::Error: The target table 'users' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.:

**Comment by saeohioalpha** [comment]
Looking through the code, it doesn't look this is supported. I think I may try and fork to add the functionality for myself. Want me to see if I can do that and submit it back to you to merge into the main project if you like it.

**Comment by metaskills** [comment]
> I also had a funky issue about a week ago with the .exists method that would do a one as one query, but would have an offset fetch next without an order by

I've seen that when people have the composite primary keys gem installed. Make sure that is not installed. There have been reports it really messes with things. See: https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/issues?utf8=✓&q=is%3Aissue%20composite%20primary%20keys

This ticket here also mentions OUTPUT INSERTED have you read this? https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/issues/396

>  I think I may try and fork to add the functionality for myself.

Please read the tickets above first... also, I don't think fork is the right word you were looking for... you simple want a freedom patch to a method in your local app. Likely an initializer right? But please check those other links and let me know what you find. If indeed you are using the composite primary keys gem, you are suffering from freedom patches :)

**Comment by saeohioalpha** [comment]
I do have that gem installed, however i think the issue is the actual SQL. On a table with a trigger, you have to do a OUTPUT INSERTED.id INTO not just OUTPUT INSERTED.id and that is the error I am getting. I don't actually understand why that is a rule in SQL Server, but it is lol. 


I added the config option:
`cattr_accessor :use_output_inserted_into, instance_accessor: false`

I modified the database_statements.rb / sql_for_insert() with:
```ruby
def sql_for_insert(sql, pk, id_value, sequence_name, binds)
  if pk.nil?
    table_name = query_requires_identity_insert?(sql)
    pk = primary_key(table_name)
  end
  sql = if pk && self.class.use_output_inserted && !database_prefix_remote_server?
      quoted_pk = SQLServer::Utils.extract_identifiers(pk).quoted
      if !self.class.use_output_inserted_into
        sql.dup.insert sql.index(/ (DEFAULT )?VALUES/), " OUTPUT INSERTED.#{quoted_pk}"
      else
        tempsql = sql.insert sql.index(/ (DEFAULT )?VALUES/), " OUTPUT INSERTED.#{quoted_pk} INTO @inserted_id"
        "DECLARE @inserted_id table (id int);#{tempsql}; SELECT * FROM @inserted_id;"
      end
    else
      "#{sql}; SELECT CAST(SCOPE_IDENTITY() AS bigint) AS Ident"
    end
    super
  end
```
Now It successfully inserts and returns the id

> 

>   ^[[1m^[[35mSQL (41.7ms)^[[0m  ^[[1m^[[35mEXEC sp_executesql N'DECLARE @inserted_id table (id int);INSERT INTO [users] ([Email], [PassHash], [First], [Last]) OUTPUT INSERTED.[id] INTO @inserted_id VALUES (@0, @1, @2, @3); SELECT * FROM @inserted_id;', N'@0 nvarchar(50), @1 nvarchar(50), @2 nvarchar(50), @3 nvarchar(50)', @0 = N'9e8c7c6d04f6440184d816aa72345e27@d.com', @1 = N'7B21E9640DC110C18E9F900EA193EB02D0A832DD', @2 = N'sample', @3 = N'user'^[[0m  [["Email", nil], ["PassHash", nil], ["First", nil], ["Last", nil]]
  ^[[1m^[[35mSQL (92.8ms)^[[0m  ^[[1m^[[36mCOMMIT TRANSACTION^[[0m
User saved with id: **716189**


I could have just set use_output_inserted = false, but I feel that's going backwards in terms of SQL Server features instead of forward.

**Comment by saeohioalpha** [comment]
Realized I didn't include the SQL error for you in the beginning either. Man im off the ball.


ActiveRecord::StatementInvalid (TinyTds::Error: The target table 'users' of the DML statement cannot have any enabled triggers if the statement contains an OUTPUT clause without INTO clause.:

**Comment by metaskills** [comment]
Closed this issue in favor of discussion in #595

**Comment by metaskills** [comment]
Released v5.1.2 with this and another fix.
